blaze/compute/sparksql.py

```diff
@@ -1,135 +1,136 @@
-    schema = discover(rdd).subshape[0]
-    columns = dshape_to_alchemy(schema)
-    for column in columns:
-        if column.name == primary_key:
-            column.primary_key = True
+@dispatch(Field, SQLContext)
+def compute_up(expr, data, **kwargs):
+    name = expr._name
+    columns = dshape_to_alchemy(expr.dshape)
+    return sa.Table(name, sa.MetaData(), *columns)
 
-    metadata = sa.MetaData()  # TODO: sync this between many tables
 
-    query = sa.Table(name, metadata, *columns)
+@curry
+def istable(db, t):
+    return (isinstance(t, Field) and isrecord(t.dshape.measure) and
+            t._child.isidentical(db))
 
-    mapping = {rdd: query}
 
-    return SparkSQLQuery(context, query, mapping)
+@dispatch(Expr, SQLContext)
+def compute_down(expr, data):
+    """ Compile a blaze expression to a sparksql expression"""
+    leaves = expr._leaves()
 
+    # make sure we only have a single leaf node
+    if len(leaves) != 1:
+        raise ValueError('Must compile from exactly one root database')
 
-@dispatch(Symbol, SchemaRDD)
-def compute_up(ts, rdd, **kwargs):
-    return make_query(rdd)
+    leaf, = leaves
 
+    # field expressions on the database are Field instances with a record
+    # measure whose immediate child is the database leaf
+    tables = pipe(expr._subterms(), filter(istable(leaf)), list)
 
-@dispatch((var, Label, std, Sort, count, nunique, Selection, mean,
-           Head, ReLabel, Distinct, ElemWise, By, any, all, sum, max,
-           min, Reduction, Projection, Field), SchemaRDD)
-def compute_up(e, rdd, **kwargs):
-    return compute_up(e, make_query(rdd), **kwargs)
+    # raise if we don't have tables in our database
+    if not tables:
+        raise ValueError('Expressions not referencing a table cannot be '
+                         'compiled')
 
+    # make new symbols for each table
+    new_leaves = [symbol(t._name, t.dshape) for t in tables]
 
-@dispatch((BinOp, Join),
-          (SparkSQLQuery, SchemaRDD),
-          (SparkSQLQuery, SchemaRDD))
-def compute_up(e, a, b, **kwargs):
-    if not isinstance(a, SparkSQLQuery):
-        a = make_query(a)
-    if not isinstance(b, SparkSQLQuery):
-        b = make_query(b)
-    return compute_up(e, a, b, **kwargs)
+    # sub them in the expression
+    expr = expr._subs(dict(zip(tables, new_leaves)))
 
+    # get sqlalchemy tables, we can't go through compute_down here as that will
+    # recurse back into this function
+    sa_tables = [compute_up(t, data) for t in tables]
 
-@dispatch((UnaryOp, Expr), SparkSQLQuery)
-def compute_up(expr, q, **kwargs):
-    scope = kwargs.pop('scope', dict())
-    scope = dict((t, q.mapping.get(data, data)) for t, data in scope.items())
+    # compute using sqlalchemy
+    scope = dict(zip(new_leaves, sa_tables))
+    query = compute(expr, scope)
 
-    q2 = compute_up(expr, q.query, scope=scope, **kwargs)
-    return SparkSQLQuery(q.context, q2, q.mapping)
+    # interpolate params
+    qs = str(literalquery(query))
```

Comment:
We'll want to try to use the `hive` dialect here if possible.

Also, at least in older versions there seemed to be several `sql` methods, some of which may have been hive focused.  Hopefully they've been merged by now, might be worth looking in to.